### PR TITLE
Add std and inlay hints

### LINKED ISSUE #7: Std and prelude support
I'm trying to add rust-analyzer-wasm to playground. Main problem here is that it doesn't support std. I ( [code](https://github.com/HKalbasi/rust-analyzer-wasm) / [demo](https://silly-einstein-6e6848.netlify.app/) ) currently added a fake std with subset of code as a simple dependency, but it has some problems:
* It doesn't detect prelude. I added `use std::prelude::v1::*` and some more thing in first line invisibly. It works but it is hacky and there may be a better way.
* It doesn't detect language features. For example `std::ops::Range { }` works but `(1..num)` type is detected as `{unknown}`

What I'm missing? Should I declare std crate special in rust-analyzer? Or some flags (I'm adding `#[lang="Range"]` but it doesn't work) in std source code? And in general what is the best way to add std to rust-analyzer-wasm?

### PR DESCRIPTION
**Pull request by hkalbasi**
I actually done this as an experiment for adding [RA to playground](https://github.com/integer32llc/rust-playground/issues/357) but I think this demo can also take advantage of this PR and by merging it, future people don't need to repeat my work.

Fix #7

This satisfies me, but it is not great. Specially it is slower than RA on vscode, and I don't know why it should be that slow. Suggestions on this (and other problems) are welcome.

### CODE REVIEW

**Comment by bjorn3** [comment]
Could you avoid checking in fake_*.rs and instead add a build step to generate them? They are really big.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -36,7 +36,11 @@ jobs:
         node-version: '15'
 
     - name: Build wasm
-      run: wasm-pack build --target web
+      run: |
+        cd rust-pack
+        cargo run
```
Comment (bjorn3) [inline .github/workflows/ci.yaml:43]: Would compiling in release mode be a net win in terms of time it takes to compile and run combined?

File: rust-pack/src/remove_function.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+
```
Comment (bjorn3) [inline rust-pack/src/remove_function.rs]: Can you please run rustfmt?

File: www/index.js [hunk 3]
```diff
@@ -124,6 +120,28 @@ monaco.languages.onLanguage(modeId, async () => {
             }
         },
     });
+    monaco.languages.registerInlayHintsProvider(modeId, {
+        async provideInlayHints(model, range, token) {
+            let hints = await state.inlay_hints();
+            return hints.map((hint) => {
+                if (hint.hint_type == 1) {
```
Comment (bjorn3) [inline www/index.js:127]: What about hint type 0?

File: www/webpack.config.js [hunk 4]
```diff
@@ -27,7 +31,7 @@ module.exports = {
     },
     plugins: [
         new HtmlWebPackPlugin({
-            title: "Rust Analyzer",
+            title: "Playground",
```
Comment (bjorn3) [inline www/webpack.config.js]: maybe
```suggestion
            title: "Rust Analyzer Playground",
```

**Comment by bjorn3** [comment]
There is a fair amount of commented out code. I think it should be removed before merging.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (hkalbasi) [reply .github/workflows/ci.yaml:43]: I didn't see difference. It is fast (=<5 seconds relative to optimizing wasm which took one minute minimum) so it doesn't matter?

File: www/index.js (hunk 3 above)
Comment (hkalbasi) [reply www/index.js:127]: RA doesn't emit type 0 (type 0 is other in lsp/monaco):

```rust
                match ih.kind {
                    InlayKind::TypeHint | InlayKind::ChainingHint => InlayHintType::Type,
                    InlayKind::ParameterHint => InlayHintType::Parameter,
                },
```

Comment (bjorn3) [reply www/index.js:127]: Could you remove that enum variant on the rust side? You can explicitly specify a discriminant to keep the numbers identical.

Comment (hkalbasi) [reply www/index.js:127]: I didn't get exactly what you want. `InlayHintType` is not meaningless and is the rust equivalent of this enum: https://microsoft.github.io/monaco-editor/api/enums/monaco.languages.inlayhintkind.html

You want to send `ih.kind` and move this `match` to JS?

Comment (bjorn3) [reply www/index.js:127]: If `InlayHintType::Other` is not handled here I think it is best to remove this variant completely and only re-introduce it once it is being used. I think that will save some debugging when it becomes used.

Comment (hkalbasi) [reply www/index.js:127]: Aha, just the variant. I thought the whole enum should be removed.

File: www/index.js [hunk 5]
```diff
@@ -229,7 +247,7 @@ monaco.languages.onLanguage(modeId, async () => {
         }
     }
 
-    monaco.languages.setTokensProvider(modeId, {
+    /*monaco.languages.setTokensProvider(modeId, {
```
Comment (bjorn3) [inline www/index.js:250]: Did you comment this out as you broke it, or for another reason?

**Review by bjorn3 — APPROVED** [review]
I didn't look at rust-pack, but the rest of the code LGTM. This is my first time looking at the code though and I don't have r+ permissions for this repo. r? @matklad ?

File: www/index.js (hunk 5 above)
Comment (hkalbasi) [reply www/index.js:250]: It is broken and for my experiment propose (adding to playground) a static highlighter was enough, so I removed it. But for "rust analyzer demo" we may want to bring it back even if it is broken and worse than static.

Break reproduce:
1. go to https://ra-wasm.netlify.app/
2. remove all the code
3. press ctrl+Z.

**Comment by matklad** [comment]
bors r+

Fixed the permissions as well, so r+ should work now.

**Comment by matklad** [comment]
bors try

**Comment by hkalbasi** [comment]
What happened exactly? I can't see bors fail details.

**Comment by lnicola** [comment]
It didn't run because the `trying` branch is missing from the workflow.

**Comment by hkalbasi** [comment]
Seems there is now a `trying` branch.

bors r=@matklad

**Comment by lnicola** [comment]
I think you need to add it to the workflow on your branch.

bors r=matklad

**Comment by lnicola** [comment]
Yeah, I don't think it works. Can you add it there?

**Comment by hkalbasi** [comment]
"it" is the bors.toml? I rebased to master if it is.

**Comment by lnicola** [comment]
No, sorry. Add the `trying` branch to the list at the top of `.github/workflows/ci.yaml`, under `on/push/branches`.

**Comment by lnicola** [comment]
bors r=matklad

**Comment by lnicola** [comment]
Nope, sorry, it's still not working.

**Comment by lnicola** [comment]
Can you also add `staging`?

**Comment by lnicola** [comment]
bors try

**Comment by lnicola** [comment]
bors r=matklad

**Comment by lnicola** [comment]
It's up on https://ra-wasm.netlify.app/, and it's awesome!

**Pull request merged**